### PR TITLE
python: support interface to perform KVS watch

### PR DESCRIPTION
**Pull request by chu11**
Built on top of #5373 and #5374.

I basically did what was done with `event_watch_async()` and `JobEventWatchFuture`.  Then I created a parent class for both to avoid code duplication.

Notes:

- I call the parent class `WatchABC`, dunno if that's the best name.  I nearly went with "WatchInterface", but knew that wasn't Pythonic wording.  I went with "ABC" (Abstract Base Class), but don't love that either.

- I left `WatchABC` in `kvs.py` which maybe isn't the best place.  Could be in a `watch.py`??  Ehh ...

- I didn't add a synchronous `kvs_watch()` function like `event_watch()`.  As I began testing and I thought about it more, there is probably less need for the function.  Unlike "eventlog watching", general "KVS watch" does not generally have a "known end" to the stream.  Without a known end, calling a synchronous `kvs_watch()` makes less sense.

- The tests I wrote are a little funky.  I didn't want to go through the process of having a thread off to the side updating KVS values, so I often made two KVS watchers, one that does updates during callbacks.

- Unfortunately `flux.kvs.commit()` is only synchronous, which made tests harder to do.  I added a hidden `_wait` flag to the function to skip blocking in the function.  Instead I could have added `flux.kvs.commit_async()` which returns a future, but I figure that might be for another day.

- Am going to try and get #4792  to work with this interface.  See if it all is good.

### CODE REVIEW

**Review by grondo — COMMENTED** [review]
Noted a couple trivial things on a first pass.

**Comment by chu11** [comment]
> Do you have a reference for WatcherInterface not being Pythonic? FWIW, we already have ResourceSetImplementation, but that is a somewhat special case since it is defining the interface that an implementation used by the ResourceSet class will use.

I think it's mostly because I'm using the "ABC" class, so I think Python uses the term "Abstract Base Class" more than "Interface" (which I believe is the term used mostly in the Java world), although I now see that you used class "ABC" In "ResourceSetImplementation".  For consistency, we could go with "Implementation" but ...

> Actually, thinking abstractly (😜), the existing JobEventWatchFuture is a specific case of a generic KVS watch, so maybe the right thing to do here is to just derive JobEventWatchFuture from the new KVSWatchFuture and therefore not need an abstract class at all.

Ahhh that's a pretty good idea, I didn't think of doing it that way.

> Actually it might be better to do that now, since it is much more useful than the hidden _wait parameter, and could work for someone trying to use the KVS asynchronously (and actually get errors from the commit)

Lemme work on that in a separate PR.

**Comment by grondo** [comment]
> Lemme work on that in a separate PR.

Well, it would remove the need for the hidden `_wait` parameter wouldn't it? Maybe best to not add then remove that parameter, unless there is something I'm not understanding.

**Comment by chu11** [comment]
> Well, it would remove the need for the hidden _wait parameter wouldn't it? Maybe best to not add then remove that parameter, unless there is something I'm not understanding.

I was thinking I would put it in a new PR to precede this one.  Then I can update this :-).

**Comment by grondo** [comment]
Oh, sounds good! Sorry I confused myself into thinking the commit was added here, but obviously that's not true. :facepalm:

**Comment by chu11** [comment]
Re-pushed basing it on PR #5627 which uses the new `KVS.commit_async()` instead of a hidden `_wait` flag.

Renamed `WatchABC` to `WatchImplementation`.  I elected to rename it rather than have `JobEventWatchFuture` based on `KVSWatchFuture`.  I forgot that the actual `get()` and `cancel()` functions are a little tricky, which we do not want `JobEventWatchFuture` to re-implement.  We only want it to implement a dumb function to (for example) call `event_watch_get()`.

File: src/bindings/python/flux/kvs.py [hunk 1]
```diff
@@ -712,3 +719,106 @@ def walk(directory, topdown=False, flux_handle=None, namespace=None):
...
+        for it to exit.  Defaults to False.
+    :param uniq: If True, only different values will be returned by
+        watch.  Defaults to False.
+    :param full: If True, any change that can affect the key is
+        monitored.  Typically, this is to capture when a parent directory
+        is removed or altered in some way.  Typically kvs watch will not
+        detect this as the exact key has not been changed.  Defaults to
+        False.
+    :returns: a KVSWatchFuture object
+    :rtype: KVSWatchFuture
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/bindings/python/flux/kvs.py:907]: Want to update this one to Google docstring format, or pass on that for now?

**Review by grondo — APPROVED** [review]
LGTM. Just one question about the docstring format for `kvs_watch_async`

**Comment by chu11** [comment]
Thanks!  Re-pushed updating the google docstring style.  Figure I better do it now or I'll forget later :P

**Pull request merged**